File: C++/Linked List/15_middle_of_linked_list.cpp
/*
Idea:
Using 2 pointer technique:
front to be moved twice at one iteration & behind to be moved once.
once front reaches end, return value of behind, if null return -1
*/

//{ Driver Code Starts
//Initial template for C++

#include <bits/stdc++.h>
using namespace std;

struct Node
{
    int data;
    struct Node* next;

    Node(int x) {
        data = x;
        next = NULL;
    }
};
void printList(Node* node)
{
    while (node != NULL) {
        cout << node->data << " ";
        node = node->next;
    }
    cout << "\n";
}


// } Driver Code Ends

class Solution {
public:
    /* Should return data of middle node. If linked list is empty, then  -1*/
    int getMiddle(Node* head)
    {
        Node* temp = head;
        Node* mid = head;

        if (head == NULL)
            return -1;

        while (mid != NULL && mid->next != NULL)
        {
            mid = mid->next->next;
            // this pointer moves 2 nodes ahead everytime loop is run

            temp = temp->next;
            // this pointer moves 1 node ahead everytime loop is run
        }

        // since temp was moving with half speed, it is there at halfway point
        return temp->data;
    }
};


//{ Driver Code Starts.

int main() {
    //code
    int t;
    cin >> t;
    while (t--) {
        int N;
        cin >> N;
        int data;
        cin >> data;
        struct Node* head = new Node(data);
        struct Node* tail = head;
        for (int i = 0; i < N - 1; ++i)
        {
            cin >> data;
            tail->next = new Node(data);
            tail = tail->next;
        }

        Solution ob;
        cout << ob.getMiddle(head) << endl;
    }
    return 0;
}

// } Driver Code Ends

File: Python/Linked List/15_middle_of_linked_list.py
"""
Idea:
Using 2 pointer technique:
front to be moved twice at one iteration & behind to be moved once.
once front reaches end, return value of behind, if null return -1
"""

# your task is to complete this function



class Solution:
    #  Should return data of middle node. If linked list is empty, then  -1
    def findMid(self, head):
        if head is None:
            return None

        ptr1 = head
        ptr2 = head
        while ptr2 is not None and ptr2.next is not None:
            ptr1 = ptr1.next
            # this pointer moves 1 node ahead everytime loop is run

            ptr2 = ptr2.next.next
            # this pointer moves 2 nodes ahead everytime loop is run

        return ptr1.data
        # since slow was moving with half speed, it is there at halfway point


# Node Class
class node:
    def __init__(self):
        self.data = None
        self.next = None


# Linked List Class
class Linked_List:
    def __init__(self):
        self.head = None
        self.tail = None

    def insert(self, data):
        if self.head == None:
            self.head = node()
            self.tail = self.head
            self.head.data = data
        else:
            new_node = node()
            new_node.data = data
            new_node.next = None
            self.tail.next = new_node
            self.tail = self.tail.next


def printlist(head):
    temp = head
    while temp is not None:
        print(temp.data, end=" ")
        temp = temp.next
    print('')


# Driver Program
if __name__ == '__main__':
    t = int(input())
    for i in range(t):
        list1 = Linked_List()
        n = int(input())
        values = list(map(int, input().strip().split()))
        for i in values:
            list1.insert(i)
        ob = Solution()
        print(ob.findMid(list1.head))

# } Driver Code Ends